File: 2021/03-binary-diagnostic/03.py
file = open("input.txt", "rt")
contents = file.read()
file.close()

lines = contents.strip().split('\n')
counts = [0] * len(lines[0])
for line in lines:
    for idx, bit in enumerate(line):
        if bit == "1":
            counts[idx] += 1

gamma = 0
epsilon = 0
for count in counts:
    gamma = gamma << 1
    epsilon = epsilon << 1
    if count < len(lines) - count:
        epsilon += 1
    else:
        gamma += 1

print(gamma * epsilon)

File: 2021/03-binary-diagnostic/03.hs
import Data.Bits
import Data.Char

main = readFile "input.txt" >>= print . solution . lines

solution :: [String] -> Int
solution xs = gamma total counts * epsilon total counts
    where counts = countAllBits xs
          total = length xs

countAllBits = foldl (zipWith (+)) (repeat 0) . map (map digitToInt)

gamma total = foldl (\acc count -> (acc `shift` 1) + (mostCommon count total)) 0
epsilon total = foldl (\acc count -> (acc `shift` 1) + (leastCommon count total)) 0

mostCommon count total | count >= total - count = 1
                       | otherwise              = 0

leastCommon count total | count < total - count = 1
                        | otherwise             = 0

